eth/tracers/api.go

```diff
@@ -916,6 +945,15 @@ func (api *API) TraceCall(ctx context.Context, args ethapi.TransactionArgs, bloc
 	}
 	defer release()
 
+	if config != nil && config.TxIndex != nil {
```

Comment:
Do we still need this? you added it to `StateAtTransaction`

Reply:
Yes but it's called conditionally above, that's why I've added it like this:

```
	if config != nil && config.TxIndex != nil {
		_, _, statedb, release, err = api.backend.StateAtTransaction(ctx, block, int(*config.TxIndex), reexec)
	} else {
		statedb, release, err = api.backend.StateAtBlock(ctx, block, reexec, nil, true, false)
	}
	if err != nil {
		return nil, err
	}
	defer release()
```

Reply:
Oh no I took the wrong condition, I need to run on the `else` clause but copied straight the `if ... ` one, will invert the logic.

Reply:
Updated, please take a look.

Reply:
> By default eth_call and debug_traceCall will simulate the call as if it's appended to the block. So they take the final state of existing block, and execute transaction on top. But because all parameters such as block number and timestamp are taken from the block itself, it is as if this call was the final transaction of that block. -> in this case we don't need to process beacon root. It has already been done.

I did not check how they retrieved their state actually, I thought the logic was to simply take state at beginning of `N+1` since it's the same as the end of `N`. All in all, I think we agree that those shouldn't have to do any beacon root processing.

> When txIdx is provided, within the same block, we go backwards up to the index. Here because we must take block N-1 and apply all the transactions to get to index the system call processing is necessary. Still block context of execution is block N.

If when `txIdx != null` the the logic is:
1. Take state at block N-1, 
1. Re-apply transactions 0 -> txIdx-1

While the context is block N, technically if your re-apply all transactions before `txIdx`, then between 1. and 2. above, beacon root must be processed. Indeed, even the first transaction of the block should see the beacon root of block N.

Is that correct?

> We should also allow users to simulate call in a potentially future block. They can already change the block context like number timestamp coinbase, etc. Now we need to add the beacon root field to BlockOverrides and if that's present do the processing too. Hmmm this creates a tricky scenario: what if I want to simulate N+3. Because the beacon root contract allows me to query all last 8192 beacon roots, I wonder how we should handle this scenario. It is sorta similar to BLOCKHASH. I believe there we default to 0x000...000 for hash of "phantom" blocks (e.g. N+2 in this case).

Of I wasn't aware it was possible to "simulate" future block. Indeed as soon as you start "to create" new blocks, the beacon root for unknown block doesn't exist. But I'm unsure about the code changes this comment should bring/remove.

Anything I need to change?

Reply:
Sorry for the rambling. I wanted to have a thorough record of all the cases that could happen. As for this PR the only change I would like to see is regarding block simulations.

Please add beacon root as a field to `internal/ethapi.BlockOverrides`. It should be processed in `eth_call` and `debug_traceCall` optionally to manipulate the beacon root of a future block in the simulation.

Reply:
I just realized technically it is possible to override the beacon root for a block through state overrides because the contract is part of the state. So let's skip that request for now.